core/ir/riscv64/instr.c

```diff
@@ -120,46 +162,40 @@ bool
 instr_is_call_arch(instr_t *instr)
 {
     int opc = instr_get_opcode(instr);
-    return (opc == OP_jal || opc == OP_jalr || opc == OP_c_j || opc == OP_c_jal ||
-            opc == OP_c_jr || opc == OP_c_jalr);
+    return ((opc == OP_jal || opc == OP_jalr || opc == OP_c_jal || opc == OP_c_jalr) &&
+            opnd_get_reg(instr_get_dst(instr, 0)) != DR_REG_ZERO);
 }
 
 bool
 instr_is_call_direct(instr_t *instr)
 {
-    /* FIXME i#3544: Check if valid */
     int opc = instr_get_opcode(instr);
-    return (opc == OP_jal);
+    return ((opc == OP_jal || opc == OP_c_jal) &&
+            opnd_get_reg(instr_get_dst(instr, 0)) != DR_REG_ZERO);
 }
 
 bool
 instr_is_near_call_direct(instr_t *instr)
 {
-    /* FIXME i#3544: Check if valid */
-    int opc = instr_get_opcode(instr);
-    return (opc == OP_jal);
+    return instr_is_call_direct(instr);
 }
 
 bool
 instr_is_call_indirect(instr_t *instr)
 {
-    /* FIXME i#3544: Check if valid */
     int opc = instr_get_opcode(instr);
-    return (opc == OP_jalr);
+    return ((opc == OP_jalr || opc == OP_c_jalr) &&
+            opnd_get_reg(instr_get_dst(instr, 0)) != DR_REG_ZERO);
 }
 
 bool
 instr_is_return(instr_t *instr)
 {
-    /* FIXME i#3544: Check if valid */
     int opc = instr_get_opcode(instr);
-    if (instr_num_srcs(instr) < 1 || instr_num_dsts(instr) < 1)
-        return false;
-    opnd_t rd = instr_get_dst(instr, 0);
-    opnd_t rs = instr_get_src(instr, 0);
-    return (opc == OP_jalr && opnd_get_reg(rd) == DR_REG_X0 &&
-            opnd_is_memory_reference(rs) && opnd_get_base(rs) == DR_REG_X1 &&
-            opnd_get_disp(rs) == 0);
+    return ((opc == OP_c_jr || opc == OP_jalr) &&
+            opnd_get_reg(instr_get_dst(instr, 0)) == DR_REG_X0 &&
```

Comment:
I assume these opcodes are guaranteed to have 2 srcs and 1 dst?

Reply:
Yes.